### PR TITLE
Breaking: Normalize error messages

### LINKED ISSUE #1133: Breaking changes
- [ ] Normalize / Standardize error massages. (See: #1170)
- [x] Make event related changes (done in #1055)
- [x] Change `getRuleName` to `getRulePath` (See: #1134)
- [x] Extract connectors to packages (@sarvaje)
- [x] Webpack bundle, we will probably break packages for the `utils` part (@molant) --> Not bundling the main CLI but rather optimize the dependencies
- [x] Extract wizards to their own packages (See: #1132) (@molant)
- [x] Remove `plugins` [from config](https://github.com/sonarwhal/sonarwhal/blob/a929058af3b44eaf89627f361f3d5c9ef31ab22d/packages/sonarwhal/src/lib/types.ts#L64) (done in #1141).
- [x] Event to evalute JS (See : #497)
- [x] Move `cdp-async-html` to `utils-debugging-protocol` package (See #1184).
- [x] Other (?)

### LINKED ISSUE #1170: Standardize error messages
Per #1133, I'm listing all the error messages we report so we can normalize them if needed.

**amp-validator**

Outputs the errors by the validator

**apple-touch icon**

| current | proposed |
| ---------| -----------| 
| 'apple-touch-icon' should have non-empty 'href' attribute | |
| '${appleTouchIconHref}' file request failed | |
| '${appleTouchIconHref}' could not be fetched (status code: ${response.statusCode}) | |
| '${appleTouchIconHref}' is not a valid PNG | |
| '${appleTouchIconHref}' is not a PNG | |
| '${appleTouchIconHref}' is not 180x180px | |
| No 'apple-touch-icon' was specified | |
| 'rel' attribute value should be 'apple-touch-icon' | |
| 'sizes' attribute is not needed | |
| 'apple-touch-icon' should be specified in the '<head>' | |
| A 'apple-touch-icon' was already specified | |

**axe**

| current | proposed |
| -------- | ---------- |
| Error executing script: "${e.message}". Please try with another connector | |

**babel-config**

It broadcasts the errors by the parser

**content-type**

| current | proposed |
| -------- | ---------- |
| 'content-type' header was not specified | |
| 'content-type' header should have the value '${userDefinedMediaType}' | |
| 'content-type' header value is invalid (${e.message}) | |
| 'content-type' header should have media type '${mediaType}' (not '${originalMediaType}') | |
| 'content-type' header should have 'charset=${charset}'${originalCharset ? ` (not '${originalCharset}')` : ''} | |
| 'content-type' header should not have 'charset=${originalCharset}' | |

**disown-opener**

| current | proposed |
| -------- | ---------- |
|  '${cutString(await element.outerHTML(), 100)}' is missing 'rel' ${requiredValues.length === 1 ? 'value' : 'values'} '${requiredValues.join('\', \'')}' | |

**highest-available-document-mode**

| current | proposed |
| -------- | ---------- |
| 'x-ua-compatible' header was not specified | |
| 'x-ua-compatible' header is not needed | |
| 'x-ua-compatible' header value should be 'ie=edge' | |
| Meta tag is not needed | |
| Meta tag usage is discouraged, use equivalent HTTP header | |
| No 'x-ua-compatible' meta tag was specified | |
| The value of 'content' should be 'ie=edge' | |
| Meta tag needs to be included before all other tags except for the '<title>' and the other '<meta>' tags | |
| Meta tag should not be specified in the '<body>' | |
| A 'x-ua-compatible' meta tag was already specified | |

**html-checker**

| current | proposed |
| -------- | ---------- |
| Couldn't get results from HTML checker for ${resource}. Error: ${error} | | 
| Output the errors from the scanner | |

**http-cache**

| current | proposed |
| -------- | ---------- |
| No "cache-control" header or empty value found. It should have a value | |
| The ${invalidDirectives.size === 1 ? 'directive' : 'directives'} ${Array.from(invalidDirectives.keys()).join(', ')} ${invalidDirectives.size === 1 ? 'is' : 'are'} invalid | |
| The following ${invalidValues.size === 1 ? 'directive has' : 'directives have'} an invalid value:\n${directivesToString(invalidValues)} | |
| The directive "${nonRecommendedDirective}" is not recommended | |
| The following Cache-Control header is using a wrong combination of directives:\n${header} | |
| The target should not be cached, or have a small "max-age" value (${maxAgeTarget}):\n${header} | |
| Static resources should have a long cache value (${maxAgeResource}) and use the immutable directive:\n${header} | |
| No configured patterns for cache busting match ${resource}. See docs to add a custom one. | |

**http-compression**

| current | proposed |
| -------- | ---------- |
| Should be served with the 'Vary' header containing 'Accept-Encoding' value. | |
| Should not be served compressed with ${encoding} as the compressed size is ${sizeDifference > 0 ? 'bigger than' : 'the same size as'} the uncompressed one. | |
| Could not be fetched when requested compressed with Brotli | |
| Should${notRequired ? ' not' : ''} be served compressed${encoding ? ` with ${encoding}` : ''}${notRequired ? '' : ` when ${['Zopfli', 'gzip'].includes(encoding) ? 'gzip' : encoding} compression is requested`}${suffix ? `${!suffix.startsWith(',') ? ' ' : ''}${suffix}` : ''}. | |
| Should${notRequired ? ' not' : ''} be served with the 'content-encoding${encoding ? `: ${encoding}` : ''}' header${suffix ? ` ${suffix}` : ''}. | |
| Could not be fetched when requested compressed with Brotli | |
| Could not be fetched when requested compressed with gzip | |
| Could not be fetched | |
| Disallowed compression method: '${encoding}'. | |
| Could not be fetched when requested uncompressed | |
|  Should not be served with the 'content-encoding' header. | |

**https-only**

| current | proposed |
| -------- | ---------- |
| The site should be HTTPS | |
| Shouldn't be redirected from HTTP | |
| Should be served over HTTPS | |`

**image-optimization-cloudinary**

| current | proposed |
| -------- | ---------- |
| No valid configuration for Cloudinary found. Hint coudn't run. | |
| File ${cutString(file.originalUrl)} could be around ${sizeDiff.toFixed(2)}kB (${percentageDiff}%) smaller. | |
| The total size savings optimizing the images in ${data.resource} could be of around ${totalSavings.toFixed(0)}kB. | | 

**manifest-app-name**

| current | proposed |
| -------- | ---------- |
| Should contain the '${propertyName}' property | |
| Should have non-empty '${propertyName}' property value | |
| Should have the '${propertyName}' property value under ${shortNameLengthLimit} characters | |

**manifest-exists**

| current | proposed |
| -------- | ---------- |
| Web app manifest not specified | |
| A web app manifest file was already specified | |
| Should have non-empty 'href' | |

**manifest-file-extension**

| current | proposed |
| -------- | ---------- |
| The file extension should be '${standardManifestFileExtension}'${fileExtension ? ` (not '${fileExtension}')` : ''} | |

**manifest-is-valid**

| current | proposed |
| -------- | ---------- |
| '${property}' property value ('${colorValue}') is invalid | |
| '${property}' property value ('${colorValue}') is not supported everywhere | |
| 'lang' property value ('${manifest.lang}') is not a valid language tag | |
| Should contain valid JSON | |

**meta-charset-utf-8**

| current | proposed |
| -------- | ---------- |
| No charset meta tag was specified | |
| Use shorter '`<meta charset="utf-8">`' | |
| The value of 'charset' is not 'utf-8' | |
| Charset meta tag should be the first thing in '`<head>`' | |
| Meta tag should not be specified in the '`<body>`' | |
| A charset meta tag was already specified | |

**meta-theme-color**

| current | proposed |
| -------- | ---------- |
| No 'theme-color' meta tag was specified | |
| 'content' attribute value ('${contentValue}') is invalid | |
| 'content' attribute value ('${contentValue}') is not supported everywhere | |
| 'name' attribute needs to be 'theme-color' (not '${nameAttributeValue}') | |
| A 'theme-color' meta tag was already specified | |
| Should not be specified in the '<body>' | |

**meta-viewport**

| current | proposed |
| -------- | ---------- |
| Meta tag should have non-empty 'content' attribute | |
| Meta tag has unknown property: '${key}' | |
| Meta tag has invalid value '${content.invalidValues[key]}' for property '${key}' | |
| Meta tag has disallowed property: '${key}' | |
| Meta tag should have 'width=device-width' | |
| Meta tag should have 'initial-scale=1' | |
| No viewport meta tag was specified | |
| Meta tag should not be specified in the '<body>' | |
| A viewport meta tag was already specified | |

**minified-js**

| current | proposed |
| -------- | ---------- |
| JavaScript content could be minified | |

**no-bom**

| current | proposed |
| -------- | ---------- |
| Error fetching the content | |
| Text based resources shouldn't start with the BOM character to force UTF-8 encoding | |

**no-broken-links**

| current | proposed |
| -------- | ---------- |
| Broken link found (domain not found)' | |
| Broken link found (${brokenStatusCodes[statusIndex]} response) | |

**no-disallowed-headers**

| current | proposed |
| -------- | ---------- |
| 'Server' header value contains more than the server name | |
| '${headers.join('\', \'')}' ${numberOfHeaders === 1 ? 'header is' : 'headers are'} disallowed | |

**no-friendly-error-pages**

| current | proposed |
| -------- | ---------- |
| Response with status code ${key} had less than ${threshold} bytes | |

**no-html-only-headers**

| current | proposed |
| -------- | ---------- |
| '${headers.join('\', \'')}' ${numberOfHeaders === 1 ? 'header is' : 'headers are'} not needed | |

**no-html-only-headers**

| current | proposed |
| -------- | ---------- |
| ${response.hops.length} ${response.hops.length === 1 ? 'redirect' : 'redirects'} detected for ${cutString(request.url)} (max is ${maxHops}). | |

**no-p3p**

| current | proposed |
| -------- | ---------- |
| P3P is deprecated and should not be used | |

**no-protocol-relative-urls**

| current | proposed |
| -------- | ---------- |
| Protocol relative URL found: ${url} | |

**no-vulnerable-javascript-libraries**

| current | proposed |
| -------- | ---------- |
| ${library.name}@${library.version} has ${vulnerabilities.length} known ${vulnerabilities.length === 1 ? 'vulnerability' : 'vulnerabilities'} (${detail}). See ${link} for more information. | |
| Error executing script: "${e.message}". Please try with another connector | |

**performance-budget**

| current | proposed |
| -------- | ---------- |
| To load all the resources on a ${config.id} network, it will take about ${loadTime.toFixed(1)}s in optimal conditions.<br>That's ${(loadTime - config.load).toFixed(1)}s more than the ${config.load}s target. | |

**sri**

| current | proposed |
| -------- | ---------- |
| Cross-origin scripts need a "crossorigin" attribute to be eligible for integrity validation | |
| Attribute "crossorigin" doesn't have a valid value, should "anonymous" or "use-credentials": crossorigin="${crossorigin}" | |
| Resource ${resource} requested without the "integrity" attribute | |
| The format of the "integrity" attribute should be "sha(256|384|512)-HASH": ${integrity.substr(0, 10)}… | |
| The hash algorithm "${algorithms[highestAlgorithmPriority]}" doesn't meet the baseline "${this.baseline}" | |
| The resource is not delivered via a secure context | |
| The hash in the "integrity" attribute doesn't match the received payload. <br>Expected: ${integrities.join(', ')} <br>Actual:   ${hashes.join(', ')} | |

**ssllabs**

| current | proposed |
| -------- | ---------- |
| ${resource} doesn't support HTTPS. | |
| ${serverName}'s grade ${grade} doesn't meet the minimum ${minimumGrade} required. | |
| Couldn't get results from SSL Labs for ${resource}. | |
| Didn't get any result for ${resource}.<br>There might be something wrong with SSL Labs servers. | |

**strict-transport-security**

| current | proposed |
| -------- | ---------- |
| Error with getting preload status for ${resource}. | |
| Error with getting preload status for ${resource}. There might be something wrong with the verification endpoint. | |
| Error with getting preload eligibility for ${resource}. | |
| 'strict-transport-security' header should't be specified in pages served over HTTP. | |
| 'strict-transport-security' header was not specified | |
| 'strict-transport-security' header requires 'max-age' directive | |
| 'strict-transport-security' header 'max-age' value should be more than ${minMaxAgeValue} | |

**stylesheet-limit**

| current | proposed |
| -------- | ---------- |
| Maximum of ${maxImports} nested imports reached (${results.imports}) | |
| Maximum of ${maxRules} CSS rules reached (${results.rules}) | |
| Maximum of ${maxSheets} stylesheets reached (${results.sheets}) | |

**typescript-config**

| current | proposed |
| -------- | ---------- |
| Couldn't find package "tslib". | |
| Based on your browser configuration your "compilerOptions.target" should be "${maxESVersion}". Current one is "${target}" | |

**validate-set-cookie-header**

| current | proposed |
| -------- | ---------- |
| '${headerName}' header to set '${setCookie.name}' has trailing ';' | |
| '${headerName}' header contains unknown attribute '${directiveKey}'. | |
| '${headerName}' header contains more than one ${directiveKey}. | |

**webpack-config**

| current | proposed |
| -------- | ---------- |
| webpack configuration file not found in your project. | |
| webpack is not installed in your project. | |
| The parser webpack-config should be activated | |
| The parser typescript-config should be activated | |
| TypeScript `compilerOptions.module` option should be `esnext` | |
| The parser babel-config should be activated | |
| Babel presets `modules` option should be `false` | |
| \`${config.devtool.toString()}\` not recommended for prodution | |

**x-content-type-options**

| current | proposed |
| -------- | ---------- | 
| 'x-content-type-options' header is not specified | |
| 'x-content-type-options' header value (${headerValue}) is invalid | |
| 'x-content-type-options' header is not needed | |

And this are the error messages for JSON validations (`schema-validator.ts`):

| type | current | proposed |
| ----- | -------- | ---------- |
| additional property | ${property ? `'${property}' ` : ''}${property ? error.message : `${error.message[0].toLocaleUpperCase()}${error.message.substr(1)}`}. Additional property found '${additionalProperty}'. | |
| enum | '${property}' ${error.message} '${allowedValues.join(', ')}'. Value found '${error.data}' | |
| pattern | '${property}' ${error.message.replace(/"/g, '\'')}. Value found '${error.data}' | |
| type | '${property}' ${error.message.replace(/"/g, '\'')}. | |

### PR DESCRIPTION
**Pull request by alrra**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)

Fix #1170
Fix #1133

### CODE REVIEW

File: packages/hint-amp-validator/src/hint.ts [hunk 1]
```diff
@@ -70,11 +73,14 @@ export default class AmpValidatorHint implements IHint {
...
-                    // We ignore errors that are not 'ERROR' if user has configured the hint like that
+                    /*
+                     * We ignore errors that are not 'ERROR'
+                     * if user has configured the hint like that.
+                     */
                     if (errorsOnly && error.severity !== 'ERROR') {
                         debug(`AMP error doesn't meet threshold for reporting`);
                     } else {
-                        await context.report(resource, null, message, null, { column: error.column, line: error.line });
+                        await context.report(resource, null, `${message}`, null, { column: error.column, line: error.line });
```
[the remark this answers is not in the document]
Comment (alrra) [reply packages/hint-amp-validator/src/hint.ts]: > why are you using an string template?

Fixed.

I must have been running some kinda macro on the file.

File: packages/hint-minified-js/tests/tests.ts [hunk 2]
```diff
@@ -3,14 +3,15 @@ import { HintTest } from '@hint/utils-tests-helpers/dist/src/hint-test-type';
...
+
 const generateScriptTag = (script) => {
     return `<script>${script}</script>`;
 };
 
-const hintPath = getHintPath(__filename);
-
-const expectedMessageFromHint = 'JavaScript content could be minified';
-const unminifiedJs= `
+const unminifiedJS= `
```
Comment (sarvaje) [inline packages/hint-minified-js/tests/tests.ts]: `unminifiedJS =`

File: packages/hint/src/lib/utils/misc/pretty-print-array.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * e.g.:
+ *
+ *  []          =>
+ *  [1]         => '1'
+ *  [1, 2]      => '1' and '2'
+ *  [1, '2', 3]   => '1', '2', and '3'
+ *
+ */
+
+export default (array: (string|number)[]): string => {
```
Comment (sarvaje) [inline packages/hint/src/lib/utils/misc/pretty-print-array.ts]: Use the format `Array<string | number>` for consistency.

File: packages/hint-minified-js/tests/tests.ts (hunk 2 above)
Comment (alrra) [reply packages/hint-minified-js/tests/tests.ts]: Fixed.

File: packages/hint/src/lib/utils/misc/pretty-print-array.ts (hunk 3 above)
Comment (alrra) [reply packages/hint/src/lib/utils/misc/pretty-print-array.ts]: Fixed.

File: packages/hint-apple-touch-icons/tests/tests.ts [hunk 4]
```diff
@@ -5,14 +5,31 @@ import { getHintPath } from 'hint/dist/src/lib/utils/hint-helpers';
...
-const imageWithIncorrectDimensions = fs.readFileSync(`${__dirname}/fixtures/incorrect-dimensions.png`); // eslint-disable-line no-sync
 const imageThatIsNotSquare = fs.readFileSync(`${__dirname}/fixtures/not-square.png`); // eslint-disable-line no-sync
+const imageWithIncorrectDimensions = fs.readFileSync(`${__dirname}/fixtures/incorrect-dimensions.png`); // eslint-disable-line no-sync
 const imageWithIncorrectFileFormat = fs.readFileSync(`${__dirname}/fixtures/incorrect-file-format.png`); // eslint-disable-line no-sync
-// const imageWithTransparentBackground = fs.readFileSync(`${__dirname}/fixtures/transparent-background.png`); // eslint-disable-line no-sync
-const appleTouchIconLinkTag = '<link rel="apple-touch-icon" href="/apple-touch-icon.png">';
+
+// Error messages
+
+const elementAlreadySpecifiedErrorMessage = `'apple-touch-icon' link element is not needed as one was already specified.`;
```
[the remark this answers is not in the document]
Comment (alrra) [reply packages/hint-apple-touch-icons/tests/tests.ts]: Yes, I will do that with some other changes in another PR, later this week.

**Pull request closed without merging**